src/mock_crust/support.rs

```diff
@@ -0,0 +1,449 @@
+        if let Some(endpoint) = self.find_endpoint_by_peer_id(peer_id) {
+            self.send_packet(endpoint, Packet::Message(data));
+            true
+        } else {
+            false
+        }
+    }
+
+    pub fn prepare_connection_info(&self, result_token: u32) {
+        // TODO: should we also simulate failure here?
+        // TODO: should we simulate asynchrony here?
+
+        let result = ConnectionInfoResult {
+            result_token: result_token,
+            result: Ok(OurConnectionInfo(self.peer_id, self.endpoint)),
+        };
+
+        self.send_event(Event::ConnectionInfoPrepared(result));
+    }
+
+    pub fn connect(&self, _our_info: OurConnectionInfo, their_info: TheirConnectionInfo) {
+        let TheirConnectionInfo(_, peer_endpoint) = their_info;
+        self.send_packet(peer_endpoint, Packet::ConnectRequest(self.peer_id));
+    }
+
+    fn send_packet(&self, receiver: Endpoint, packet: Packet) {
+        self.network.send(self.endpoint, receiver, packet);
+    }
+
+    fn receive_packet(&mut self, sender: Endpoint, packet: Packet) {
+        // TODO: filter packets
+
+        match packet {
+            Packet::BootstrapRequest(peer_id) => self.handle_bootstrap_request(sender, peer_id),
+            Packet::BootstrapSuccess(peer_id) => self.handle_bootstrap_success(sender, peer_id),
+            Packet::BootstrapFailure => self.handle_bootstrap_failure(sender),
+            Packet::ConnectRequest(peer_id) => self.handle_connect_request(sender, peer_id),
+            Packet::ConnectSuccess(peer_id) => self.handle_connect_success(sender, peer_id),
+            Packet::ConnectFailure(peer_id) => self.handle_connect_failure(sender, peer_id),
+            Packet::Message(data) => self.handle_message(sender, data),
+            Packet::Disconnect => self.handle_disconnect(sender),
+        }
+    }
+
+    fn handle_bootstrap_request(&mut self, peer_endpoint: Endpoint, peer_id: PeerId) {
+        if self.is_listening() {
+            self.handle_bootstrap_accept(peer_endpoint, peer_id);
+            self.send_packet(peer_endpoint, Packet::BootstrapSuccess(self.peer_id));
+        } else {
+            self.send_packet(peer_endpoint, Packet::BootstrapFailure);
+        }
+    }
+
+    fn handle_bootstrap_accept(&mut self, peer_endpoint: Endpoint, peer_id: PeerId) {
+        self.add_connection(peer_id, peer_endpoint);
+        self.send_event(Event::BootstrapAccept(peer_id));
+    }
+
+    fn handle_bootstrap_success(&mut self, peer_endpoint: Endpoint, peer_id: PeerId) {
+        self.add_connection(peer_id, peer_endpoint);
+        self.send_event(Event::BootstrapConnect(peer_id));
+        self.decrement_pending_bootstraps();
+    }
+
+    fn handle_bootstrap_failure(&mut self, _peer_endpoint: Endpoint) {
+        self.decrement_pending_bootstraps();
+    }
+
+    fn handle_connect_request(&mut self, peer_endpoint: Endpoint, peer_id: PeerId) {
+        // TODO: ignore if already connected?
```

Comment:
I would rather log a warning in that case, so that our tests let us know if Routing is attempting duplicate connections.